File: algorithms/leet.0190.src.1.c
uint32_t reverseBits(uint32_t n) {
    // swap a and b:
    // a = a ^ b
    // b = a ^ b
    // a = a ^ b
    n ^= (0x0000FFFF & n) << 16;
    n ^= (0xFFFF0000 & n) >> 16;
    n ^= (0x0000FFFF & n) << 16;
    n ^= (0x00FF00FF & n) << 8 ;
    n ^= (0xFF00FF00 & n) >> 8 ;
    n ^= (0x00FF00FF & n) << 8 ;
    n ^= (0x0F0F0F0F & n) << 4 ;
    n ^= (0xF0F0F0F0 & n) >> 4 ;
    n ^= (0x0F0F0F0F & n) << 4 ;
    n ^= (0x33333333 & n) << 2 ;
    n ^= (0xCCCCCCCC & n) >> 2 ;
    n ^= (0x33333333 & n) << 2 ;
    n ^= (0x55555555 & n) << 1 ;
    n ^= (0xAAAAAAAA & n) >> 1 ;
    n ^= (0x55555555 & n) << 1 ;
    return n;
}

File: algorithms/leet.0190.src.1.py
class Solution {
public:
    uint32_t reverseBits(uint32_t n) {
        n = ((n>>16)&0x0000FFFF) | ((n<<16)&0xFFFF0000);
        n = ((n>>8)&0x00FF00FF) | ((n<<8)&0xFF00FF00);
        n = ((n>>4)&0x0F0F0F0F) | ((n<<4)&0xF0F0F0F0);
        n = ((n>>2)&0x33333333) | ((n<<2)&0xCCCCCCCC);
        n = ((n>>1)&0x55555555) | ((n<<1)&0xAAAAAAAA);
        return n;
    }
};
